feat(GUI): move drive selector warning to flash step

Shou opened the pull request:
We move the drive selector warning to the flash step, and concatenate
warning messages when more than one needs to be displayed at once.

Change-Type: patch
Changelog-Entry: Move the drive selector warning dialog to the flash step.

lurch commented on lib/gui/components/drive-selector/controllers/drive-selector.js:
```diff
@@ -56,6 +57,20 @@ module.exports = function (
   this.drives = availableDrives
 
+  /**
+   * @summary 
```
Should this have more complete docstrings?

lurch commented on lib/gui/components/drive-selector/controllers/drive-selector.js:
```diff
@@ -75,20 +90,38 @@ module.exports = function (
-    if (constraints.isDriveSizeRecommended(drive, selectionState.getImage())) {
-      return $q.resolve(true)
+    const isLargeDrive = constraints.isDriveSizeLarge(drive)
+    const isRecommendedDrive = constraints.isDriveSizeRecommended(drive, selectionState.getImage())
```
Minor nit, but could we rename this variable to `isRecommendedSizeDrive` ?
To me, a variable named `isRecommendedDrive` implies that this drive is recommended above all other drives :)

Shou commented:
Screenshot of concatenated warnings

![image](https://user-images.githubusercontent.com/819413/33896130-fad113bc-df59-11e7-988d-dfa76f31c44a.png)

There's a lot of space above when the text is long enough, looking into fixing that.

lurch commented on lib/gui/components/drive-selector/controllers/drive-selector.js:
```diff
@@ -75,20 +90,38 @@ module.exports = function (
+        ]
+      ]
+
+      const concatenatedMessage = _.reduce(messageTuples, (accumulatedMessage, messageTuple) => {
```
Do we really need to concatenate the messages? Seems _extremely_ unlikely that a drive will be both 'large' and 'smaller than the recommended size' ;-)

lurch commented on lib/gui/components/drive-selector/controllers/drive-selector.js:
```diff
@@ -75,20 +90,38 @@ module.exports = function (
+          } else {
+            return accumulatedMessage
+          }
+        }, '')
```
Why does this need to use `reduce` rather than just a straightforward `join`?

Shou replied:
I guess we could `filter` and then `join` but that's two loops, but the reason we use `reduce` here is so we can pick which strings to add to the dialog in one recursion over the list rather than two since JavaScript doesn't do stream-fusion. This doesn't matter anymore now that we're not using this though!

lurch replied:
Ahhh, I'd misunderstood what `messageTuples` actually contains - didn't realise it contained a boolean condition flag in addition to the message itself.

lurch commented on lib/shared/store.js:
```diff
@@ -144,6 +144,9 @@ const storeReducer = (state = DEFAULT_STATE, action) => {
 
+          // We don't want to auto-select large drives
+          !constraints.isDriveSizeLarge(drive),
+
```
Should the unit-tests be added to? ;-)

lurch commented on lib/shared/messages.js:
```diff
@@ -67,6 +67,11 @@ module.exports = {
+
+    largeDriveSize: _.template([
+      'It looks like you\'re trying to flash a large drive of size <%= humanDriveSize %>,',
+      'this could potentially be a backup drive containing important data.',
```
Hmmm, not sure if we specifically want to mention "backup drive" here? Although I like the "important data" phrasing.
How about something like:
"It looks like you're trying to flash a large drive with size 120 GB, are you sure this disk doesn't contain important data? Etcher will overwrite everything."
?
I suspect @alexandrosm will want some input here :)

jviotti replied:
We can simplify this to:

> Because of its size, this drive doesn't look like a removable drive. Are you sure you want to flash it?

jviotti replied:
I don't think we need to mention that 64 GB is the limit that triggers the heuristic

Shou replied:
I think we should mention the size too, and we could also add the drive `description`. It's very helpful to the user in identifying which drive they've chosen just from that message since it covers the drive selector. What about:

> Because of its size of <%= humanDriveSize %>, <%= driveDescription %> doesn't look like a removable drive. Are you sure you want to flash it?

lurch replied:
I think the "removable drive" language isn't what we want either though, because somebody was complaining about accidentally overwriting their portable backup drive using Etcher - and in this case it actually _is_ a removable drive.

lurch replied:
What about something like:

> <%= driveDescription %> is quite large (<%= humanDriveSize %>). Please check it doesn't contain any important data before Etcher overwrites it.

lurch commented on lib/shared/drive-constraints.js:
```diff
@@ -278,6 +278,23 @@ exports.isDriveSizeRecommended = (drive, image) => {
+/**
+ * @summary 64GB
+ */
+const LARGE_DRIVE_SIZE = 64000000000
```
I wonder if this should be base-2 units i.e. 68719476736 ?

jhermsmeier replied:
Storage devices usually have sizes in base-10, so base-2 wouldn't trigger on a 64GB SD-card, for example. Guess that depends on when we want that to happen.

lurch replied:
Yeah, I mentioned in flowdock that we probably _don't_ want this triggering on 64 GB SD cards, as I suspect they'll continue to become more prevalent as their price continues to drop. In my extensive experience, SD cards are always some-number-of-megabytes smaller than their stated capacity, so no 64GB card has a capacity of 65,536 MB.
For the record, the "64 GB" SD cards I have have exact byte-capacities of 62587404288, 62730010624, 62881005568, 63416827904, 64289243136. (even the biggest one has a capacity of 'only' 61,331 MB)

And I've never seen a 64GB hard drive, they always seem to be a "round" number of units: 80GB, 100GB, 2TB, etc.

lurch replied:
OTOH if we _do_ want this to trigger for 64 GB SD cards too, then it would probably make sense (for similar reasons as above) to set this to the base-2 32GB size, i.e. 34359738368 and this would then also catch 40 GB hard drives but no "32 GB" SD cards.

jviotti commented:
I think the "recommended size" part of the message can go away. We should try to be brief rather than throwing to paragraphs at the user.

Shou commented:
Let me know what you think now

![image](https://user-images.githubusercontent.com/819413/33915123-8c98f91a-df99-11e7-84b6-97eec51ec05f.png)

lurch commented:
Do you think it's obvious enough that clicking on the grey 'X' in the top-right of the "Attention" dialog is how a (possibly nervous?) user answers "No" to the question being asked of them?  Also looks a bit weird with Etcher displaying two dialogs at once, which I guess @konmouz might want to comment on?

lurch commented on lib/shared/drive-constraints.js:
```diff
@@ -288,8 +290,12 @@ const LARGE_DRIVE_SIZE = 64000000000
+ *
+ * @example
+ * if (constraints.isDriveSizeLarge(drive)) {
+ *   console.log('Impressive')
```
:laughing:

lurch commented on lib/gui/components/drive-selector/controllers/drive-selector.js:
```diff
@@ -90,35 +97,18 @@ module.exports = function (
-        }, '')
+    if (constraints.isDriveSizeLarge(drive)) {
+      return confirmationWarningModal(messages.warning.largeDriveSize({
+        humanDriveSize: units.bytesToClosestUnit(drive.size),
```
Hmmm, borrowing the 'pattern' from the display-units PR, I wonder if it _might_ make sense for Etcher to automatically add a `humanSize` property to every drivelist object, to prevent this `units.bytesToClosestUnit` being done multiple times in different places?

lurch commented:
Oh, looking at the code in more detail, it seems that this warning dialog would actually pop up **as soon as** they click on the drive in the drive-selection dialog? Is that the UX that we want? I wonder if it might be better to only display the dialog when they actually click on the "Continue" button on the drive-selection dialog, so that the drive-dialog has now disappeared? And if the "large drive dialog" then had a "No" button, clicking that could unselect the drive that they'd chosen. (IYSWIM)

Sorry if I'm sounding a bit picky, but I don't want us to make a repeat of the accessibility-with-orange-borders UX fiasco ;-)

Shou commented:
@lurch if having several visible modals open is undesirable we don't necessarily have to close the drive selector to hide it while the other warning modal is open – we could use CSS to hide it in a generic way such that only the most recent modal is shown at a time. I believe this would work:

```scss
.modal[aria-hidden="true"] {
  display: none;
}
```

lurch commented:
I'm still not sure if displaying the modal automatically as soon as they click on the drive is desirable though? Feels a bit "in your face" to me?
This is obviously a UX question for @konmouz

jviotti commented:
I was thinking we should show it when the user clicks "Flash!", as a last opportunity to halt the process.

lurch commented:
> as a last opportunity to halt the process.

If we're doing that, then IMHO we _definitely_ need a far more obvious "No" button ;-)

lurch commented:
And BTW GitHub is now annoyingly auto-hiding the conversation (on `lib/shared/messages.js`) about what the warning message actually _is_ - I still think it needs some tweaks.

Shou commented:
This is definitely a good change for when multi-selection and writes arrive as well. If the user cancels the warning dialog we should probably re-open the drive selector as that's the obvious step anyway?

konmouz commented:
I agree with @lurch, it would be better to have 2 CTA's, one for 'continue' and one for 'cancel'. We don't need the 'x' button in this attention modal. Also, WRT the warning on large drive size selection, we can follow the flow of a system drive. The suspicious drive can get a label, and then when selected the 'continue' button becomes red and also gets an alert just above it. Finally, you should get another attention/confirmation modal after clicking the 'flash' CTA.

Shou commented:
So we want to move the warnings (not recommended + large drive) to appear when clicking flash in this PR?

jviotti commented:
I think we should try to open the least amount of modals as possible, and treat this as an exceptional case to protect the user from OSes that think that external HDDs are not fixed drives.

> If the user cancels the warning dialog we should probably re-open the drive selector as that's the obvious step anyway, i.e. they want to change the selected drive(s)?

Sounds reasonable. So the plan is the following:

- User selects a drive larger than 64 GB
- User clicks Flash
- Since the drive is larger than 64 GB, we show the warning modal
- If the user agrees, then everything carries on
- If the user doesn't agree, then the warning modal closes and the drive selector modal opens instead

konmouz commented:
So this is the behaviour when selecting a system drive:

<img width="810" alt="screen shot 2017-12-14 at 10 43 26" src="https://user-images.githubusercontent.com/15323961/33988240-b7960450-e0bb-11e7-9572-0afe3836fe7d.png">
<img width="813" alt="screen shot 2017-12-14 at 10 43 35" src="https://user-images.githubusercontent.com/15323961/33988242-b7eedb52-e0bb-11e7-9716-379212c410dc.png">
<img width="807" alt="screen shot 2017-12-14 at 10 43 54" src="https://user-images.githubusercontent.com/15323961/33988244-b8101bb4-e0bb-11e7-81af-4223fb9de135.png">

lurch commented:
@konmouz There's two separate issues here. One is system drives (i.e. internal fixed drives, which we only display when 'unsafe mode' is enabled) and the other is very large (i.e. over 64GB) external drives.
Based on @konmouz 's comments above, do we also want a `LARGE DRIVE` warning-label displayed in the drive-selector window?

konmouz commented:
@lurch, I believe so. `LARGE DRIVE`s should follow the `SYSTEM DRIVE`s behaviour.

Shou commented:
Let me know what you think now

![image](https://user-images.githubusercontent.com/819413/34050482-f8f401e8-e1b2-11e7-800b-9344015e6f90.png)
![image](https://user-images.githubusercontent.com/819413/34050493-0269dc98-e1b3-11e7-8d87-acf73a416e41.png)
![image](https://user-images.githubusercontent.com/819413/34050503-0d5dd442-e1b3-11e7-9eb7-f7eb5382cdde.png)

lurch commented:
Looks good, but I still think the message needs tweaking. In some earlier comments (which GitHub is now auto-hiding), I said:

I think the "removable drive" language isn't what we want either though, because somebody was complaining about accidentally overwriting their portable backup drive using Etcher - and in this case it actually _is_ a removable drive.

What about something like:

> <%= driveDescription %> is very large (<%= humanDriveSize %>). Please check it doesn't contain any important data before Etcher overwrites it.

lurch commented:
Looks like you need to rebase / update your shrinkwrap file too.

jviotti commented on lib/shared/drive-constraints.js:
```diff
@@ -428,3 +465,7 @@ exports.getDriveImageCompatibilityStatuses = (drive, image) => {
   return statusList
 }
+
+exports.hasDriveImageCompatibilityStatus = (drive, image) => {
```
Missing JSDoc?

alexandrosm commented:
How about "Drive <...> is unusually large for an SD card or USB stick. Are
you sure you want to continue?"

-- 

*Alexandros Marinos*

Founder & CEO, Resin.io


+1 206-637-5498

@alexandrosm

On Fri, Dec 15, 2017 at 12:25 PM, Juan Cruz Viotti <notifications@github.com
> wrote:

> *@jviotti* commented on this pull request.
> ------------------------------
>
> In lib/shared/drive-constraints.js
> <https://github.com/resin-io/etcher/pull/1917#discussion_r157291816>:
>
> > @@ -428,3 +465,7 @@ exports.getDriveImageCompatibilityStatuses = (drive, image) => {
>
>    return statusList
>  }
> +
> +exports.hasDriveImageCompatibilityStatus = (drive, image) => {
>
> Missing JSDoc?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/resin-io/etcher/pull/1917#pullrequestreview-83916807>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABLUCBQM8zcvp8ZIokcuf4K4wVKT8ni6ks5tAtWkgaJpZM4Q_QVQ>
> .
>

lurch commented on lib/gui/pages/main/controllers/flash.js:
```diff
@@ -16,19 +16,98 @@
+  const displayTailoredWarning = (drive, image) => {
+    const messageKeys = {
+      largeDriveSize: constraints.isDriveSizeLarge(drive),
+      unrecommendedDriveSize: !constraints.isDriveSizeRecommended(drive, image)
```
Hmm, I'm not sure if unrecommendedDriveSize is a serious-enough-problem to warrant displaying an extra warning modal to the user?

Shou replied:
We could remove it but it's the current behaviour, so if we want to remove it I think we should open a separate issue.

lurch commented on lib/gui/pages/main/controllers/flash.js:
```diff
@@ -16,19 +16,98 @@
+      }
+    }
+    const messageKey = _.findKey(messageKeys, _.identity)
+    const message = messages.warning[messageKey](messageArgs[messageKey])
```
IMHO this line looks a little funny, with 4 different variables with `message` in the name, and one of the variables being used in two different places to index into two different objects.

I _think_ I see what you're doing with using `const` and lodash all over the place, but I wonder if something like this might be a little bit more readable?
```js
let message = null
if (constraints.isDriveSizeLarge(drive)) {
    message = messages.warning[largeDriveSize]({humanDriveSize: units.bytesToClosestUnit(drive.size), drive})
} else if (!constraints.isDriveSizeRecommended(drive, image)) {
    message = messages.warning[unrecommendedDriveSize]({image, drive})
}
```
?

Or if you really wanted to keep the const-ness I guess you _could_ do:
```js
const message = constraints.isDriveSizeLarge(drive)
    ? messages.warning[largeDriveSize]({humanDriveSize: units.bytesToClosestUnit(drive.size), drive})
    : (
        !constraints.isDriveSizeRecommended(drive, image)
        ? messages.warning[unrecommendedDriveSize]({image, drive})
        : null
    )
```
but that obviously starts heading into the unreadable-code category ;-)

Or perhaps you could do something like:
```js
const messageValues = [
    [constraints.isDriveSizeLarge(drive), messages.warning[largeDriveSize]({humanDriveSize: units.bytesToClosestUnit(drive.size), drive})],
    [!constraints.isDriveSizeRecommended(drive, image), messages.warning[unrecommendedDriveSize]({image, drive})]
}
// and then use `_.find` etc.
```
which is more similar to the approach you're already using, but gets rid of the separate `messageArgs` variable?

\*shrug\* I dunno really, @jviotti can probably offer better advice when it comes to Javascript best-practices ;-)  Feel free to ignore me.

Shou replied:
I noticed a bug with my current implementation here and now I think the best option is to just go for the if-then-else

lurch commented on lib/shared/drive-constraints.js:
```diff
@@ -405,6 +437,11 @@ exports.getDriveImageCompatibilityStatuses = (drive, image) => {
       type: exports.COMPATIBILITY_STATUS_TYPES.ERROR,
       message: exports.COMPATIBILITY_STATUS_MESSAGES.LOCKED
     })
+  } else if (exports.isDriveSizeLarge(drive)) {
```
Should the unit-tests that deal with `getDriveImageCompatibilityStatuses` be updated too?

Shou replied:
Done

lurch commented on lib/shared/drive-constraints.js:
```diff
@@ -405,6 +437,11 @@ exports.getDriveImageCompatibilityStatuses = (drive, image) => {
+      type: exports.COMPATIBILITY_STATUS_TYPES.WARNING,
+      message: exports.COMPATIBILITY_STATUS_MESSAGES.LARGE_DRIVE
+    })
   } else if (!_.isNil(drive) && !_.isNil(drive.size) && !exports.isDriveLargeEnough(drive, image)) {
```
Hmmm, should `isDriveSizeLarge` (which is a warning) be checked for _after_ `isDriveLargeEnough` (which is an error) ?

Hmmm number 2, if we have a drive which is both 'large' and 'system', I wonder if we should display both warning labels?

Shou replied:
That's true, we should considering we do the same with unrecommended drive/image size.

Shou commented:
I've rebased and also added [some changes][1]. Let me know what you think now. 

[1]: https://github.com/resin-io/etcher/compare/5ea1c1eac11cb8521af949247b56bafdbaaca1c9...warn-large-drives

jhermsmeier commented:
Conflicts need resolving

jhermsmeier commented:
@Shou can you add an updated screenshot to the PR description?

Shou commented:
@jhermsmeier I've split off part of this PR into https://github.com/resin-io/etcher/pull/2014 and will probably split off some more soon

Shou commented:
Split off into https://github.com/resin-io/etcher/pull/2045 – this PR will be about moving the warning dialog to the flash step instead, because they don't depend on each other so it'll be faster to get the PRs through independently/concurrently.

Shou commented:
Current screenshot

![image](https://user-images.githubusercontent.com/819413/38268324-b47c79e6-3775-11e8-8962-8b34a1b8acf9.png)

jhermsmeier commented on lib/gui/app/pages/main/controllers/flash.js:
```diff
@@ -81,6 +158,11 @@ module.exports = function (
       }
     }).catch((error) => {
+      // When flashing is cancelled before starting above there is no error
+      if (!error) {
```
I'm going to add this to the list of problems one wouldn't have without Promises ;P

jhermsmeier commented on lib/gui/app/pages/main/controllers/flash.js:
```diff
@@ -25,19 +26,88 @@ const exceptionReporter = require('../../../modules/exception-reporter')
    *
    * @param {Object} image - image
-   * @param {Array<Object>} drives - drives
+   * @param {Array<Object>} devices - drive devices
```
The type doc here needs updating to `Array<String>` too I think

Shou replied:
Done that now

jhermsmeier commented on lib/gui/app/pages/main/controllers/flash.js:
```diff
@@ -25,19 +26,88 @@ const exceptionReporter = require('../../../modules/exception-reporter')
+   * @function
+   * @public
+   *
+   * @param {Array<Object>} drives - list of drive objects
```
The type doc here needs updating to `Array<String>` I believe

Shou replied:
It requires `Object` in order for the `constraints` predicates to work

jhermsmeier replied:
Oh, right - missed that
